### DISCUSSION TITLE
Spec for `meltano config` support for encryption

### DISCUSSION THREAD

**Discussion by edgarrmondragon** [meltano/meltano#6999] [Spec Discussions]
Part of a multi-part set of feature releases:

1. https://github.com/meltano/meltano/discussions/6999 _(this item)_
   - 1. https://github.com/meltano/meltano/issues/6987
1. https://github.com/meltano/meltano/discussions/7055
1. https://github.com/meltano/meltano/discussions/7056

----------------------

The CLI will support encryption of config values.

The following command will encrypt `<config value>`

```
meltano config <plugin name> set --secret <config name> <config value>
```

## Automatic secret detection

All settings of kind `password` will automatically encrypted when set in the CLI, if there's a `default_encryption_key`:

```console
$ meltano config <a plugin> set <a password setting> <a secret>
```

## Storing encrypted secrets

Secrets will be stored in a structured string similar to [Sops's](https://github.com/mozilla/sops) notation:

```
ENC(name=kms,uri='awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1',cipher=...)
```

The exact format of this string is still TBD, but it would at least encode:

- The name of the key _used_ for encryption
- The `uri` of the key _needed_ for decrytion
- The secret ciphertext

Secrets (their ciphertext) are stored along with metadata that helps Meltano identify the decryption key.

```yaml
plugins:
  extractors:
  - name: <a plugin>
    config:
      <a password setting>: 'ENC(name=kms,uri='awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1',cipher=...)'
```

The original encryption key is stored inline to ensure the secret can always be decrypted, even if keys are removed from the project at some point.

## Encryption and Decryption

1. All keys are identified by a unique ID or name presented.
1. There's always ONE AND ONLY ONE encryption key (EK) enabled in a project, referenced by name as in `secrets_config.meltano-cloud`.
1. There's ONE OR MORE decryption keys (DK) in the `secrets_config.keys[]` array.

The corresponding _ciphertext_ will be processed by

- a key managed by Meltano Cloud
- A bring-your-own key with initial support for AWA KMS, and with plans to support HashiCorp Vault and GCP KMS:
  
  ```yaml
  secrets_config:
    keys:
    - name: kms
      uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
    - name: vault
      uri: hashivault://mykey 
    - name: meltano-cloud
      uri: meltano://project-id
    encryption_key: meltano-cloud
    default_store: dotenv  # not needed for V1
  ```

## Key rotation

Not supported in V1.

Changelog:

- 2022-11-18: Changed `secrets_provider` to `default_encryption_key`

cc @aaronsteers

**Comment by edgarrmondragon** [comment]
@aaronsteers I've updated the spec based on our conversation in Slack

**Reply by aaronsteers** [reply]
Looks great! Thanks, @edgarrmondragon !

**Reply by aaronsteers** [reply]
Attn: @meltano/engineering @meltano/product for spec review ☝️

**Reply by tayloramurphy** [reply]
@aaronsteers @edgarrmondragon the provided spec looks good. 

The remaining questions I have are about the next level on this:

* What do the logs say if there isn't a default_encryption key?
* Do we prompt the user if they're setting a password w/o a `default_encryption_key`?

Another thing I just thought of - do we want to have a somewhat interactive component on this? I can imagine folks may not want to have the secret in their `history` on the command line - should we accept `meltano config <a plugin> set <a password setting>` and then prompt for the value?

**Reply by aaronsteers** [reply]
Re:

> Another thing I just thought of - do we want to have a somewhat interactive component on this? I can imagine folks may not want to have the secret in their history on the command line - should we accept meltano config <a plugin> set <a password setting> and then prompt for the value?

Logged this new issue to address:

- https://github.com/meltano/meltano/issues/7003

I think we can tackle together or separate, but it's definitely worthwhile in its own right. And they do stack nicely, I think. 😄

**Reply by aaronsteers** [reply]
@edgarrmondragon - Do you want to address these points:

> The remaining questions I have are about the next level on this:
>
> - What do the logs say if there isn't a default_encryption key?
> - Do we prompt the user if they're setting a password w/o a default_encryption_key?

**Reply by edgarrmondragon** [reply]
@tayloramurphy @aaronsteers wdyt about a warning message when setting passwords without a key? One thing I worry about this approach is the interaction with alternative setting stores (environments, `.env`, etc.)

**Reply by tayloramurphy** [reply]
@edgarrmondragon that's a good call out. I see in the spec we have:

> All settings of kind password will automatically encrypted when set in the CLI, if there's a `default_encryption_key`

Are we essentially updating the behavior such that we wouldn't allow the writing of passwords _at all_ to meltano.yml unless they're encrypted? Otherwise if it's going to `.env` or the systemdb it wouldn't have to be - that's, presumably, when we could alert the user about the option to encrypt. 

I think that would work if I've articulated it correctly.

**Reply by aaronsteers** [reply]
> Otherwise if it's going to .env or the systemdb it wouldn't have to be - that's, presumably, when we could alert the user about the option to encrypt.

I think a warning in this case would make sense. 

> Setting saved successfully to system DB: (redacted)
> Warning: The setting value was stored in clear text because `--encryption-key` was not specified and no `default_encryption_key` was available in project settings. For more information, please see https://melta.no#encryption

wdyt, @edgarrmondragon ?

**Reply by edgarrmondragon** [reply]
@aaronsteers sounds good. So to make sure I understand the interaction with stores, do the following scenarios make sense?

1. set password setting, without any encryption configured, store is determined automatically as today and `meltano.yml` is _never_ used, warning is printed

1. set password setting, with encryption configured, the "normal" store priority is used, the password is set in `meltano.yml`

1. set non password setting, with encryption configured, setting is not encrypted, warning printed that setting will not be encrypted since it's not a password, setting is stored as currently

**Reply by aaronsteers** [reply]
Great question! I wasn't trying to propose a different behavior for autostore, but it is someone we should consider.

With values being encrypted, stores like meltano.yml could be viable when otherwise they'd not be preferred.

My sense is that probably the auto-store logic should still work as today, so that secrets still land in _more_ secure locations, but we might promote in our docs that users can choose (if they want) to commit back their encrypted secrets to yaml when/if they want to have those available on other runners or for other developers.

Wdyt?

**Reply by aaronsteers** [reply]
Regarding non-secrets, I think this is probably reasonable:

1. If encryption provider is specified in yaml but not in the CLI (like a port or a boolean value, for instance), them we don't encrypt. (Encryption provider exists, but no need to encrypt.)
2. If encryption flag or --secret is specified in the CLI (for a host address, for instance, that _might_ be sensitive), then the value is encrypted. (We basically override the settings' default sensitive level when the user requests it.)

TBD if the auto store in the second case should follow the store decision tree for secrets or non-secrets . Since it's encrypted, we could make the case that storing in yaml is fine, but since the user can excited request yaml storage if they want it, I might slightly lean towards following the decision tree for secrets in those cases, just for extra safety.

**Reply by edgarrmondragon** [reply]
> stores like meltano.yml could be viable when otherwise they'd not be preferred.

IIUC, that's the main value add of at-rest encryption such as we're discussing, right?

**Reply by aaronsteers** [reply]
> > stores like meltano.yml could be viable when otherwise they'd not be preferred.
> 
> IIUC, that's the main value add of at-rest encryption such as we're discussing, right?

Definitely. Encrypted values will be and to be stored safely where otherwise they could not be. I'm just not yet sure if store location should be automatically changed from how at have it, or if store override should still be opt-in.

For example, taking Meltano Cloud out of the picture, I'm sure many users would like to apply at-rest encryption when saving config values to postgres or .env. 

Just to say, the decision to add encryption may or may not be paired with a decision to change where values are stored.

**Comment by edgarrmondragon** [comment]
_Let's discuss key rotation here_

Another aspect worth considering for V1 is key rotation.

Suppose a user has been encrypting secrets with their key for some time, then someone (presumably a security team) rotates their key. What happens to their encrypted secrets then?

To rotate the key, all encrypted secrets would have to be reencrypted with the new key. So support for 2 or more keys is required, the first key is used for encryption while _all_ the keys are tried in order for decryption, failing if none is able to decrypt the values.

For a concrete example, one of the most basic forms of symmetric encryption is Fernet. This is how [Airflow manages Fernet key rotation](https://airflow.apache.org/docs/apache-airflow/stable/security/secrets/fernet.html):

<blockquote>
<h2>Rotating encryption keys<a title="Permalink to this heading"></a></h2>
<p>Once connection credentials and variables have been encrypted using a fernet
key, changing the key will cause decryption of existing credentials to fail. To
rotate the fernet key without invalidating existing encrypted values, prepend
the new key to the <code>fernet_key</code> setting, run
<code>airflow rotate-fernet-key</code>, and then drop the original key from
<code>fernet_key</code>:</p>
<ol>
<li><p>Set <code>fernet_key</code> to <code>new_fernet_key,old_fernet_key</code></p></li>
<li><p>Run <code>airflow rotate-fernet-key</code> to re-encrypt existing credentials with the new fernet key</p></li>
<li><p>Set <code>fernet_key</code> to <code>new_fernet_key</code></p></li>
</ol>
</blockquote>

SOPS doesn't let you add a new key without explicit requesting for a rotation: https://github.com/mozilla/sops#262command-line

cc @aaronsteers

**Reply by aaronsteers** [reply]
Thanks for raising! This was one reason to keep a distinction between 'secrets backend' (out of scope) and 'encryption backend' (in scope) and also distinguishing 'default' encryption key that applies _currently_ either to the project or environment - vs the specific key that was applied at a specific point in time for the particular setting value in question.

1. I do think we need to handle the above cases you mention.
2. We should allow for cases where the encryption key provided at the project/environment level is not the same as the default key.
3. During decryption, we have to check metadata to see if we have access to the key that was actually used for encryption. We can't assumer the project-level setting is _currently_ relevant to every single encrypted value in the project.

If the decryption engine is literally a key file, then presumably while two keys are overlapping in lifecycle, some secrets will be able to be decrypted with one or the other key, and our metadata should point not only to the algorigthm or backend, but also to the fingerprint or id of the key we actually need.


If the decryption engine is a 3rd party service, like KMS or Meltano Cloud, then we either need the decryption service to be smart enough to figure out the right decription key from context _or_ we need to send another specific token or fingerprint value to make sure that while keys are being rotated upstream, we still have the right facilities in place to decrypt the value during the overlap period.

Does this work? Wdyt?

**Reply by aaronsteers** [reply]
After mulling on this for another 24 hours, I am wondering if the spec itself can encompass an explicit override to the default settings backend when dealing with things that are secrets? I'm wondering if the overall spect (but not necessarily first implementation) should be also considering overrides to the store location. To keep these bucketed together, we could pull both the default encryption key and also an optional default secret store into the same umbrella, something like `secrets_config` or similar.

What this looks like if adding the encryption key options and a new `default_store` (or similar) under a top-level header:

```yml
secrets_config:
  default_encryption_key:
    uri: meltano://project-id
    # uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
    # uri: hashivault://mykey
  default_store: dotenv
  # default_store: systemdb
  # default_store: kms://...
```

What's nice about thinking about both together, even if we don't build both in the first iteration, is that I think it better addresses your questions and concerns above.

Here are some theoreticals:

Use encryption key from `meltano login` to encrypt secrets directly to the yaml file:

```yml
secrets_config:
  default_encryption_key:
    uri: meltano://project-id
  default_store: meltano.yml
```

Use KMS to encrypt settings before writing to Postgres.

```yml
secrets_config:
  default_encryption_key:
    uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
  default_store: systemdb
```

Use a secret store that has built-in encryption, so no need to pre-encrypt:

```yml
secrets_config:
  default_store: vault://...
```

These could be side-by-side without a `secrets_config` umbrella:

```yml
default_secret_store: systemdb
default_encryption_key:
  uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
```

Or with a more parallel structure to allow additional config params:

```yml
default_secret_store:
  uri: systemdb
default_encryption_key:
  uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
```

Presumably, the secret store would be limited to setting stores that Meltano already supports. Even without new store options, the combination of the encryption key and also a presumed storage location provides a less ambiguous behavior for how the project designer wants to deal with `kind: password` settings and/or when using `--secret`.

What do you think? 🤔

**Comment by edgarrmondragon** [comment]
How about the following requirements:

1. All keys are identified by a unique ID or name.
2. There's always ONE AND ONLY ONE _encryption key_ (EK) enabled in a project
3. There's ONE OR MORE _decryption keys_ (DK)

Point 1 ensures that Meltano always knows which key to use for encryption.

Point 2 ensures that keys can be rotated. To do so, users

1. change their EK to a DK,
2. replace with the new key as the sole EK,
3. rotate keys, perhaps with a `meltano config rotate-keys` command that decrypts everything with the appropriate DK and re-encrypts it with the unique EK
4. remove the deprecated DK

Users don't wouldn't even need to execute steps 3 and 4 while the lifetimes of the new EK and the now-legacy DK overlap.

> we have to check metadata to see if we have access to the key that was actually used for encryption
> ...we either need the decryption service to be smart enough to figure out the right decription key from context or we need to send another specific token or fingerprint value to make sure that while keys are being rotated upstream

Yeah, so perhaps the used EK needs to be identified by ID in the metadata? Using your proposed nested project setting:

```yaml
secrets_config:
  keys:
  - name: kms
    uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
  - name: vault
    uri: hashivault://mykey 
  - name: meltano-cloud
    uri: meltano://project-id
  encryption_key: meltano-cloud
  default_store: dotenv  # not needed for V1
```

Then some pseudocode to parse metadata from the ciphertext string:

```python
cipher_metadata = parse_metadata("ENC(id=kms,cipher=...)")
assert key_metadata.id == "kms"
assert key_metadata.cipher = "..."

plaintext = decrypt_cipher(key_metadata)
```

@aaronsteers wdyt?

**Reply by aaronsteers** [reply]
@edgarrmondragon - I like this a lot! I think the distinction between (active) encryption key and the total collection of known/available decryption keys is very helpful. 

> ```
> secrets_config:
>   keys:
>   - name: kms
>     uri: awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1
>   - name: vault
>     uri: hashivault://mykey 
>   - name: meltano-cloud
>     uri: meltano://project-id
>   encryption_key: meltano-cloud
>   default_store: dotenv  # not needed for V1
> ```

👍

**Reply by aaronsteers** [reply]
In your example, above, where the key has a name of `kms` and uri of `awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1`, what do you think of including in our metadata "envelope" both the name and the uri?:

```
ENC(name=kms,uri='awskms://1234abcd-12ab-34cd-56ef-1234567890ab?region=us-east-1',cipher=...)
```

In this way, if we find a value encrypted and stored in a remote database (like postgres, for instance), we could still locate the key for decryption even if the reference to that key has been since removed or renamed in yaml (of course, _if and only if_ the key is accessible by its uri).

The name is a helpful description of the key's purpose and source, but if tomorrow there's a new default key called 'kms', we'd still have the info to find the key when aliases have changed/rotated. 

Also - if keys and names are indexed/inventoried in yaml, its value possible we'll have multiple versions of those keys defined (or not defined) when we consider the multiple branches and PRs that can be open at any point in time. Putting the uri in the metadata envelop (within `ENC(...)` or whatever our convention is) would ensure the runner can always decode the value _if_ it has runtime access to use that key.

Wdyt?

**Comment by aaronsteers** [comment]
@edgarrmondragon - Do we have a logged issue for this? Do you mind linking in the description if so?

**Reply by edgarrmondragon** [reply]
@aaronsteers done :)
